Change MCP tool definitions to use a handler

Azquelt opened the pull request:
This is preparatory work for supporting tools registered programatically without an annotated method.

- Add classes from quarkus-mcp: FeatureManager, ToolManager
- Split out the fields from ToolMetadata which are needed to call the method into MethodMetadata
- Create handlers for calling a bean method: SyncBeanMethodHandler, AsyncBeanMethodHandler. Much of the code here is moved from McpServlet.
- Move the Jsonb instance into McpCdiExtension since it's now needed to create the method handler at startup
- Update McpServlet.callTool to call the tool handlers
- Remove requestId from CancellationImpl as it was unused
- Make McpToolCallParams return a map of arguments rather than an array
- Fix areas of the code which were using the Tool annotation directly
- Update test utilities which create ToolMetadata

Also noticed an issue that we're not correctly handling Async tool business exceptions, so I've added tests and fixed that.

Part of #33351 
Fixes #33513

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Azquelt committed:
Change MCP tool definitions to use a handler

This is preparatory work for supporting tools registered
programatically without an annotated method.

- Add classes from quarkus-mcp: FeatureManager, ToolManager
- Split out the fields from ToolMetadata which are needed to call the
  method into MethodMetadata
- Create handlers for calling a bean method: SyncBeanMethodHandler,
  AsyncBeanMethodHandler. Much of the code here is moved from
  McpServlet.
- Move the Jsonb instance into McpCdiExtension since it's now needed to
  the method handler at startup
- Update McpServlet.callTool to call the tool handlers
- Remove requestId from CancellationImpl as it was unused
- Make McpToolCallParams return a map of arguments rather than an array
- Fix areas of the code which were using the Tool annotation directly
- Update test utilities which create ToolMetadata

martindrozdz reviewed (dismissed):
Looks good. I agree that more reviews would be beneficial as there are people with different ideas on what could be beneficial.

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/features/FeatureManager.java:
```diff
@@ -0,0 +1,146 @@
+
+        /**
+         * @return the name of the respective server configuration
+         * @see McpServer
```
Doesn't look like we're using this class `McpServer` maybe this comment needs updating

Azquelt replied:
This is a good point, we don't have the `McpServer` concept at all at the moment. I think I'll just remove this method.

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/features/FeatureManager.java:
```diff
@@ -0,0 +1,146 @@
+//
+//        RawMessage rawMessage();
+//
+//        Meta meta();
```
You might want to remove the commented features if we're not implementing them yet

Azquelt replied:
I'd rather leave them in at the moment, it's a good reminder of the things from Quarkus that we don't implement, and since they're commented out like this they won't appear in the API or the Javadoc.

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/tools/AsyncBeanMethodHandler.java:
```diff
@@ -0,0 +1,92 @@
+        try {
+            Object beanInstance = bm.getReference(method.bean(), method.bean().getBeanClass(), cc);
+
+            // Call the tool method synchronously
```
Did you mean to say async in this comment?

Azquelt replied:
I'm... not actually sure. We do actually call the method immediately on the current thread.

I think we can conclude though that the comment as it is now isn't too helpful. I think I'll replace it with one that just points out that this is where we call the method.

ttmunyunguma replied:
Makes sense

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java:
```diff
@@ -73,10 +66,15 @@ public class McpServlet extends HttpServlet {
     McpRequestTracker requestTracker;
 
+    @Inject
+    McpCdiExtension extension;
```
I would call this cdiExtention or mcpCdiExtention, to make reading the code later on more intuitive, for example when we do `jsonb = extension.getJsonb();` I feel like it would make more sense reading `jsonb = cdiExtension.getJsonb();` instead

ttmunyunguma commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java:
```diff
@@ -195,184 +193,55 @@ private void callTool(McpTransport transport) {
                     throws IllegalAccessException, IllegalArgumentException {
-        CreationalContext<Object> cc = bm.createCreationalContext(null);
 
+        ToolArguments toolArgs = createToolArguments(params.getArguments(jsonb));
```
Looks like this bit does not care whether the tool is async or not, I think it's better for it to be in the callTool method and pass the args as parameters here and also in the callToolAsync method. Similarly the registering of an ongoing transaction below, we could do that once and not repeat it here and for the async stuff

Azquelt replied:
The `createToolArguments` part probably makes sense to move out. It does mean we need to pass more arguments but that's probably ok.

I don't want to move the registering of an ongoing transaction out because the cleanup for that is handled differently and I want to be able to easily see where it's registered and where it's deregistered. However, putting the deregistration in a `cleanup` method does hide it a bit, so I may adjust it a bit to see if I can make it clearer.

ttmunyunguma reviewed (dismissed):
This looks great. This will be much more extensible with the pluggable handlers

Azquelt commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/tools/AsyncBeanMethodHandler.java:
```diff
@@ -0,0 +1,92 @@
+
+            return methodStage.thenApply(result -> createSuccessfulResponse(result))
+                              .exceptionally(throwable -> {
+                                  if (throwable instanceof InvocationTargetException) {
+                                      throwable = throwable.getCause();
+                                      if (isBusinessException(throwable)) {
+                                          return createBusinessErrorResponse(throwable);
+                                      } else {
+                                          return createNonBusinessErrorResponse(throwable);
+                                      }
+                                  } else {
+                                      return createNonBusinessErrorResponse(throwable);
+                                  }
+                              })
+                              .whenComplete((result, throwable) -> releaseCc(cc));
```
This exception handling is not correct.

`InvocationTargetException` is only thrown when calling a method by reflection, whereas here we're examining an exception found in a failed `CompletionStage`.

This also means we're missing a test for a tool method which returns a `CompletionStage` where a stage failed with a business exception. I'll add that to this PR.

Azquelt committed:
Add MCP Async Tool error tests

Test all forms of error reporting from MCP Async tools.

Azquelt committed:
Fix MCP Async tool exception handling

Handle exceptions correctly when the method returns a CompletionStage
which fails, rather than throwing the exception directly.

Azquelt commented:
I've addressed the comments, updated the Async tool error handling and added some tests. This is ready for re-review.

ttmunyunguma commented:
Looks good

Azquelt commented:
Rebased to pick up latest changes.

Only actual change is to configure the `Jsonb` instance where we create it. Previously this was done in `McpServlet`, this PR needs to move that code to `McpCdiExtension`.
